eth/catalyst/api.go

```diff
@@ -171,64 +174,87 @@ func (api *ConsensusAPI) makeEnv(parent *types.Block, header *types.Header) (*bl
-	data, ok := api.preparedBlocks[int(PayloadID)]
+func (api *ConsensusAPI) GetPayloadV1(PayloadID hexutil.Bytes) (*ExecutableDataV1, error) {
+	hash := []byte(PayloadID)
+	id := binary.BigEndian.Uint64(hash[:8])
+	data, ok := api.preparedBlocks[id]
 	if !ok {
 		return nil, &UnknownPayload
 	}
 	return data, nil
 }
 
-// ConsensusValidated is called to mark a block as valid, so
-// that data that is no longer needed can be removed.
-func (api *ConsensusAPI) ConsensusValidated(params ConsensusValidatedParams) error {
-	switch params.Status {
-	case VALID.Status:
-		return nil
-	case INVALID.Status:
-		// TODO (MariusVanDerWijden) delete the block from the bc
-		return nil
-	default:
-		return errors.New("invalid params.status")
+func (api *ConsensusAPI) ForkchoiceUpdatedV1(heads ForkchoiceStateV1, PayloadAttributes *PayloadAttributesV1) (ForkChoiceResponse, error) {
+	var emptyHash = common.Hash{}
+	if !bytes.Equal(heads.HeadBlockHash[:], emptyHash[:]) {
+		if err := api.checkTerminalTotalDifficulty(heads.HeadBlockHash); err != nil {
+			if block := api.eth.BlockChain().GetBlockByHash(heads.HeadBlockHash); block == nil {
+				// TODO (MariusVanDerWijden) trigger sync
+				return SYNCING, nil
+			}
+			return INVALID, err
+		}
+		// If the finalized block is set, check if it is in our blockchain
+		if !bytes.Equal(heads.FinalizedBlockHash[:], emptyHash[:]) {
+			if block := api.eth.BlockChain().GetBlockByHash(heads.FinalizedBlockHash); block == nil {
+				// TODO (MariusVanDerWijden) trigger sync
+				return SYNCING, nil
+			}
+		}
+		// SetHead
+		if err := api.setHead(heads.HeadBlockHash); err != nil {
+			return INVALID, err
+		}
+		// Assemble block (if needed)
+		if PayloadAttributes != nil {
+			data, err := api.assembleBlock(heads.HeadBlockHash, PayloadAttributes)
+			if err != nil {
+				return INVALID, err
+			}
+			hash := computePayloadId(heads.HeadBlockHash, PayloadAttributes)
+			id := binary.BigEndian.Uint64(hash)
+			api.preparedBlocks[id] = data
+			log.Info("Created payload", "payloadid", id)
+			// TODO (MariusVanDerWijden) do something with the payloadID?
+			hex := hexutil.Bytes(hash)
+			return ForkChoiceResponse{Status: SUCCESS.Status, PayloadID: &hex}, nil
+		}
 	}
+	return ForkChoiceResponse{Status: SUCCESS.Status, PayloadID: nil}, nil
 }
 
-func (api *ConsensusAPI) ForkchoiceUpdated(params ForkChoiceParams) error {
-	var emptyHash = common.Hash{}
-	if !bytes.Equal(params.HeadBlockHash[:], emptyHash[:]) {
-		if err := api.checkTerminalTotalDifficulty(params.HeadBlockHash); err != nil {
-			return err
-		}
-		return api.setHead(params.HeadBlockHash)
+func computePayloadId(headBlockHash common.Hash, params *PayloadAttributesV1) []byte {
+	input := make([]byte, 32+8+32+20)
+	copy(input, headBlockHash[:])
+	binary.BigEndian.PutUint64(input[32:], params.Timestamp)
+	copy(input[40:], params.Random[:])
+	copy(input[72:], params.FeeRecipient[:])
+	hash := sha256.Sum256(input)
+	return hash[:8]
```

Comment:
I think you can just instantiate a new keccak256 sponge, then write the stuff into it. like, 
```
hasher.Write(params.Random[:))
hasher.Write(params.Feerecipient[:])
...
```
Makes the code a bit more robust and simple